packages/loopring_v3.js/src/types.ts

```diff
@@ -0,0 +1,188 @@
+import BN = require("bn.js");
+import { SparseMerkleTree } from "./SparseMerkleTree";
+
+export enum BlockType {
+    RING_SETTLEMENT = 0,
+    DEPOSIT,
+    ONCHAIN_WITHDRAWAL,
+    OFFCHAIN_WITHDRAWAL,
+    ORDER_CANCELLATION
+  }
+
+  export enum BlockState {
+    NEW = 0,
+    COMMITTED,
+    VERIFIED,
+    FINALIZED,
+  }
+
+  export enum ForgeMode {
+    AUTO_UPGRADABLE = 0,
+    MANUAL_UPGRADABLE,
+    PROXIED,
+    NATIVE,
+  }
+
+  export interface Block {
+    blockIdx: number;
+
+    blockType: BlockType;
+    blockSize: number;
+    blockVersion: number;
+    data: string;
+    offchainData: string;
+
+    operator: string;
+
+    blockState: BlockState;
+
+    merkleRoot: string;
+
+    committedTimestamp: number;
+    verifiedTimestamp?: number;
+    finalizedTimestamp?: number;
+
+    numRequestsProcessed: number;
+
+    totalNumRequestsProcessed: number;
+    totalNumDepositsProccesed: number;
+    totalNumOnchainWithdrawalsProcessed: number;
+    totalNumTradesProccesed: number;
+    totalNumOffchainWithdrawalsProcessed: number;
+    totalNumOrderCancellationsProcessed: number;
+
+    transactionHash: string;
+  }
+
+  export interface Token {
+    tokenID: number;
+    address: string;
+    enabled: boolean;
+  }
+
+  export interface Deposit {
+    blockIdx?: number;
+    requestIdx?: number;
+
+    depositIdx: number;
```

Comment:
`requestIdx` is the index of the request in the list of ALL requests that were processed in all blocks, `depositIdx` is the on-chain index in the deposit queue. So if on an exchange there were 4 deposits and 4 trades than the last trade would have `requestIdx == 7` because it's the 8th request that was processed. Currently I store all requests in a single list (so also perhaps a single database table) which contains all requests of all types (deposits, withdrawals, trades,...) in the order they were processed by blocks. You can get them with `getProcessedRequests(startIdx: number, count: number)` (or `getRequestsInBlock(blockIdx: number)` per block).